### PR TITLE
Optimizes `icon_exists()` only because Lemon asked me to

### PR DESCRIPTION
**Pull request by Kapu1178**
## About The Pull Request
Every `icon_exists()` call will cache the entire file. Past me didn't realise *why* file opts were so expensive, but I do now. This is immeasurably slower on a single call, and *significantly* faster on subsequent calls to the same file.

There's a genuine argument to be made for caching all `rsc` DMIs during set up, but I don't think the tech is there yet.
## Why It's Good For The Game
Speed!

### CODE REVIEW

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -1459,23 +1459,22 @@ GLOBAL_LIST_EMPTY(transformation_animation_objects)
...
 
-	if(state in states)
-		icon_states_cache[file][state] = TRUE
-		return TRUE
-	else
-		icon_states_cache[file][state] = FALSE
-		if(scream)
-			stack_trace("Icon Lookup for state: [state] in file [file] failed.")
+	if(isnull(icon_states_cache[file][state]))
+		if(isnull(screams[file]) && scream)
```
Comment (LemonInTheDark) [inline code/__HELPERS/icons.dm:1473]: I still think this is nicer as composition, but eh

**Review by LemonInTheDark — APPROVED** [review]
thank you

File: code/__HELPERS/icons.dm [hunk 2]
```diff
@@ -1459,23 +1459,22 @@ GLOBAL_LIST_EMPTY(transformation_animation_objects)
 
 ///Checks if the given iconstate exists in the given file, caching the result. Setting scream to TRUE will print a stack trace ONCE.
 /proc/icon_exists(file, state, scream)
+	var/static/list/screams = list()
```
Comment (Mothblocks) [inline code/__HELPERS/icons.dm:1462]: Real name

Comment (Mothblocks) [reply code/__HELPERS/icons.dm:1462]: Actually remove it enitrely

File: code/__HELPERS/icons.dm (hunk 1 above)
Comment (Mothblocks) [reply code/__HELPERS/icons.dm:1473]: Yes icon_exists definitely should not be doing any stack tracing

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request closed without merging**